### PR TITLE
feat(commands): add buisness logic

### PR DESCRIPTION
**Pull request by georgettica**
the command will:
- extract the internalid
- extract the infraid
- find stopped instances
- correlate with stopped instance events

also:
- change client names to be similar across code
- add mockgen as some commands were not generated

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -30,6 +37,10 @@ build:
 test:
 	go test -race -mod=readonly ./...
 
+.PHONY: generate
+generate: bin/mockgen
```
Comment (georgettica) [inline Makefile:41]: should I add this to default actions we run? if so where

File: pkg/commands/isrunning/isrunning.go [hunk 2]
```diff
@@ -0,0 +1,85 @@
+package isrunning
```
Comment (georgettica) [inline pkg/commands/isrunning/isrunning.go:1]: I know the name is bad, but naming for me is hard so I just did what I though was right

File: pkg/commands/isrunning/isrunning.go [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+type IsRunning struct {
+	IsRunningService
+}
+
+// isUserAllowedToStop will verify if a user is allowed to stop instances.
+// as this is the thing that might change the most it is at the top.
+// additionally it is private as I don't see anyone using this outside of AreAllInstancesRunning
+func isUserAllowedToStop(username string) bool {
+	return false
+}
```
Comment (georgettica) [inline pkg/commands/isrunning/isrunning.go:43]: @rafael-azevedo we discussed this before and you said there was a quick and easy way to checking this. feel free to link it us

File: pkg/commands/isrunning/isrunning.go [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+
+	if len(stoppedInstancesEvents) == 0 {
+		return false, fmt.Errorf("the are stopped instances but no stoppedInstancesEvents, this means the instances were stopped too long ago or CloudTrail is not up to date")
+	}
+
+	for _, event := range stoppedInstancesEvents {
+		if !isUserAllowedToStop(*event.Username) {
+			return false, nil
+		}
+	}
```
Comment (georgettica) [inline pkg/commands/isrunning/isrunning.go:82]: this method is very clean but doesn't do some simple things:
- check if multiple stop events have happened to a similar instance
- check if some stopped instances don't have a correlating event
- what about the amount of instances in the cluster + the amount of running

File: pkg/commands/isrunning/isrunning_test.go [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+			clusterDeployment hivev1.ClusterDeployment
+			infraID           string
+			instance          ec2.Instance
+		)
+		BeforeEach(func() {
+			mockCtrl = gomock.NewController(GinkgoT())
+			mockClient = mock.NewMockIsRunningService(mockCtrl)
+			isRunning = isrunning.IsRunning{IsRunningService: mockClient}
+			var err error
+			cluster, err = v1.NewCluster().ID("12345").Build()
```
Comment (georgettica) [inline pkg/commands/isrunning/isrunning_test.go:37]: when I tried to merge these two lines I got an issue, this is the solution so far and I am glad it works :)

File: pkg/commands/isrunning/isrunning_test.go [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+			})
+		})
+
+		When("there user provided cannot stop the instance", func() {
+			It("should succeed are report that the not all instances are running", func() {
+				// Arrange
+				mockClient.EXPECT().GetClusterInfo(gomock.Any()).Return(cluster, nil)
+				mockClient.EXPECT().GetClusterDeployment(gomock.Eq(cluster.ID())).Return(&clusterDeployment, nil)
+				mockClient.EXPECT().ListStoppedInstances(gomock.Eq(infraID)).Return([]*ec2.Instance{&instance}, nil)
+				event := cloudtrail.Event{Username: aws.String("12345")}
```
Comment (georgettica) [inline pkg/commands/isrunning/isrunning_test.go:121]: once I have another test with this event, I will move it into the `vars` section

File: pkg/commands/isrunning/isrunning.go [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+	infraID := cd.Spec.ClusterMetadata.InfraID
+
+	stoppedInstances, err := i.ListStoppedInstances(infraID)
+	if err != nil {
+		return false, fmt.Errorf("could not retrieve instances for %s: %w", infraID, err)
+	}
+
+	// TODO: this statement is not fully what I wanted, but makes the code liner. will come back to it later
+	if len(stoppedInstances) == 0 {
```
Comment (ramonbutter) [inline pkg/commands/isrunning/isrunning.go:65]: what If we test the opposite? For the case there are sufficient masters and infra nodes, we can assume that the responsibility lies in our hands

Comment (georgettica) [reply pkg/commands/isrunning/isrunning.go:65]: for that we need to pull the amount of control plane and compute nodes

Comment (georgettica) [reply pkg/commands/isrunning/isrunning.go:78]: @ramonbutter answered your comment. I hope it's better now :)

**Review by T0MASD — COMMENTED** [review]
Very useful feature :100:! I left a comment on checking the node metadata with instance metadata.

File: pkg/commands/isrunning/isrunning.go [hunk 8]
```diff
@@ -56,9 +57,21 @@ func (i IsRunning) AreAllInstancesRunning(externalID string) (bool, error) {
 	}
 	infraID := cd.Spec.ClusterMetadata.InfraID
 
+	// TODO: we are pulling ListRunningInstances and ListStoppedInstances, if we want to save on API calls, we can call ListInstances and filter from there
+	// nodeCount is only pulled here as only here we are verifying the amount of running instances
+	nodeCount := cluster.Nodes().Total()
```
[the remark this answers is not in the document]
Comment (georgettica) [reply pkg/commands/isrunning/isrunning.go:62]: I can guide you through what I have so far (I hope my code is fairly straightforward and you can extract additional checks if you wish)

File: pkg/commands/isrunning/isrunning.go (hunk 2 above)
Comment (ramonbutter) [reply pkg/commands/isrunning/isrunning.go:1]: chgm?

Comment (georgettica) [reply pkg/commands/isrunning/isrunning.go:1]: Let's do that :)

File: pkg/commands/isrunning/isrunning.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (georgettica) [reply pkg/commands/isrunning/isrunning.go]: as discussed offline, the checks you are suggesting require `kubectl` access. CAD does not assume this access exists and chooses to verify without access to it.

maybe in a later milestone we will verify based on this

**Comment by georgettica** [comment]
I know what the error is and I fixed it, no need to go into the logs to check

File: pkg/commands/chgm/chgm.go [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+	// 2. running the command "aws iam list-users --query 'Users[?starts_with(UserName,`<INFRA_ID>`)].UserName'"
+	// 3. trimming the infra id from the front and the uuid from the back
+	operatorIamNames := []string{
+		"cloud-credential-operator-iam-ro",
+		"cloud-ingress-operator-credentials-aw",
+		"instance",
+		"managed-velero-operator-iam-credentia",
+		"openshift-image-registry",
+		"openshift-ingress",
+		"openshift-machine-api-aws",
```
Comment (ramonbutter) [inline pkg/commands/chgm/chgm.go:57]: When we are using the actual defined operator IAM here and not limit it to a generic regex like 
`case strings.HasPrefix(username, infraID)`
, it may make sense to reduce the list to the ones we know are actually allowed to do here

File: Makefile (hunk 1 above)
Comment (ramonbutter) [reply Makefile:41]: I wouldn't do it as we are committing mocks. But doesn't matter too much, you can add it if you want

File: pkg/commands/chgm/chgm.go [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+		return false, fmt.Errorf("could not retrieve cluster info for %s: %w", externalID, err)
+	}
+	id := cluster.ID()
+
+	cd, err := i.GetClusterDeployment(id)
+	if err != nil {
+		return false, fmt.Errorf("could not retrieve Cluster Deployment for %s: %w", id, err)
+	}
+	infraID := cd.Spec.ClusterMetadata.InfraID
+
```
Comment (ramonbutter) [inline pkg/commands/chgm/chgm.go:96]: just as a reminder once you got the internal ClusterID you have to do the assume role chain at that point. As we are having our base account credentials, and need to get into the customer account.
something like:
`
supportRoleArn := ocm.GetSupportRoleARN(id)
client, err = i.awsClient.AssumeRole(jumprole)    // fetched from config
ccsClient, err =  client.AssumeRole(supportRoleArn)
`

Comment (georgettica) [reply pkg/commands/chgm/chgm.go:96]: is that something I need to take care of, or should be provided by the person creating the AwsClient?

Comment (ramonbutter) [reply pkg/commands/chgm/chgm.go:96]: its ok for now, we can refactor it at one point. I think it is just required for solving that ticket. But yes I would put it somewhere else. But It has to come after the internalID was fetched which is now the beginning of your function. As this becomes very long, we can maybe put it immediately after we build up the client for the base/main account. somewhere after this line: https://github.com/openshift/configuration-anomaly-detection/blob/932ab869d1114f9b4867794a53fc2b4fb31c651c/cadctl/cmd/cluster-missing/cluster-missing.go#L41

File: pkg/commands/chgm/chgm.go [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+	// TODO: we are pulling ListRunningInstances and ListStoppedInstances, if we want to save on API calls, we can call ListInstances and filter from there
+	// nodeCount is only pulled here as only here we are verifying the amount of running instances
+	nodeCount := cluster.Nodes().Total()
```
Comment (ramonbutter) [inline pkg/commands/chgm/chgm.go:99]: can we really rely on that the information is up to date? Are cluster auto-scaling actions updating this value? 
Otherwise, is maybe answering the following question enough for us at that point?:

_" was any master or infra node been scaled down by a not authorized user "_

Meaning someone else touched SRE managed control-plane components and thus SLA is broken. Which doesn't mean we should gather as much more information as we can.

File: pkg/commands/chgm/chgm.go [hunk 12]
```diff
@@ -0,0 +1,137 @@
...
+	stoppedInstances, err := i.ListStoppedInstances(infraID)
+	if err != nil {
+		return false, fmt.Errorf("could not retrieve stopped instances for %s: %w", infraID, err)
+	}
+
+	// TODO: this statement is not fully what I wanted, but makes the code liner. will come back to it later
+	if len(stoppedInstances) == 0 {
+		return true, nil
+	}
+	fmt.Printf("the stoppedInstances are %v\n", stoppedInstances)
```
Comment (ramonbutter) [inline pkg/commands/chgm/chgm.go:118]:
```suggestion
	fmt.Printf("the stoppedInstances are %#v\n", stoppedInstances)
```

File: pkg/commands/chgm/chgm.go [hunk 13]
```diff
@@ -0,0 +1,137 @@
...
+	if err != nil {
+		return false, fmt.Errorf("could not retrieve stopped instance event: %w", err)
+	}
+
+	if len(stoppedInstancesEvents) == 0 {
+		return false, fmt.Errorf("the are stopped instances but no stoppedInstancesEvents, this means the instances were stopped too long ago or CloudTrail is not up to date")
+	}
+
+	for _, event := range stoppedInstancesEvents {
+		fmt.Printf("the event is %v\n", event)
```
Comment (ramonbutter) [inline pkg/commands/chgm/chgm.go:130]:
```suggestion
		fmt.Printf("the event is %#v\n", event)
```

File: pkg/commands/chgm/chgm.go (hunk 11 above)
Comment (ramonbutter) [reply pkg/commands/chgm/chgm.go:99]: but a mistake by the customer done once would always appear in cloud trail, and CAD would always assign the blame to the customer for all subsequent CHGM... not easy to get this right

Comment (georgettica) [reply pkg/commands/chgm/chgm.go:99]: we might need to take the stopped instance, search for a matching event in the timeframe to correlate to it

File: cadctl/cmd/cluster-missing/cluster-missing.go [hunk 14]
```diff
@@ -29,45 +30,67 @@ import (
...
+		return fmt.Errorf("could not verify all instances were running: %w", err)
+	}
+	fmt.Printf("yay! the result is %#v", res)
+
+	return nil
 }
 
 var (
-	incidentID string
+	externalClusterID string = ""
```
Comment (shibumi) [inline cadctl/cmd/cluster-missing/cluster-missing.go:85]: You can change this t o `externalClusterID string`. The initialization value of a string is the empty string. No need to set this.

File: pkg/commands/chgm/chgm.go [hunk 15]
```diff
@@ -0,0 +1,147 @@
...
+	}
+	for _, operatorIamName := range operatorIamNames {
+		// I added an additional '-' in the end to make sure it's the same as close as I can to the aws username (cannot guess the uuid)
+		operatorIamName = fmt.Sprintf("%s-%s-", infraID, operatorIamName)
+		if strings.HasPrefix(username, operatorIamName) {
+			return true
+		}
+	}
+
+	switch {
```
Comment (shibumi) [inline pkg/commands/chgm/chgm.go:77]: I would do an if out of this:

```go
if strings.HasPrefix(username, "osdManagedAdmin-") {
    return true
}
return false
```

It's easier to read and "flatter".

File: cadctl/cmd/cluster-missing/cluster-missing.go (hunk 14 above)
Comment (georgettica) [reply cadctl/cmd/cluster-missing/cluster-missing.go:85]: It's just so it's clear this is the default value :)

File: pkg/commands/chgm/chgm.go (hunk 15 above)
Comment (georgettica) [reply pkg/commands/chgm/chgm.go:77]: Yeah, switch case was because I had more cases.
Will change to your way

Comment (shibumi) [reply pkg/commands/chgm/chgm.go:77]: Ah okay. In that case a switch case would be the case to choose :D :D :D

File: cadctl/cmd/cluster-missing/cluster-missing.go (hunk 14 above)
Comment (georgettica) [reply cadctl/cmd/cluster-missing/cluster-missing.go:86]: Added a comment, if you like it a can resolve

Comment (shibumi) [reply cadctl/cmd/cluster-missing/cluster-missing.go:86]: typo: `s/deafault/default/`

File: pkg/commands/chgm/chgm.go [hunk 16]
```diff
@@ -74,12 +74,8 @@ func isUserAllowedToStop(username, infraID string) bool {
...
 	}
 
-	switch {
-	case strings.HasPrefix(username, "osdManagedAdmin-"):
-		return true
-	default:
-		return false
-	}
+	// I wanted to keep this an if statement, but golangci-lint didn't allow me :(
+	return strings.HasPrefix(username, "osdManagedAdmin-")
```
Comment (shibumi) [inline pkg/commands/chgm/chgm.go:78]: That's fine too

File: cadctl/cmd/cluster-missing/cluster-missing.go (hunk 14 above)
Comment (georgettica) [reply cadctl/cmd/cluster-missing/cluster-missing.go:86]: K :)

File: pkg/commands/chgm/chgm.go [hunk 17]
```diff
@@ -0,0 +1,190 @@
...
+	ListRunningInstances(infraId string) ([]*ec2.Instance, error)
+	ListStoppedInstances(infraId string) ([]*ec2.Instance, error)
+	// OCM
+	GetClusterDeployment(clusterID string) (*hivev1.ClusterDeployment, error)
+	GetClusterInfo(identifier string) (*v1.Cluster, error)
+}
+
+// Client is an implementation of the Interface, and adds functionality above it
+// to create 'Client' you can use a mock, or fill it with
+type Client struct {
```
Comment (rafael-azevedo) [inline pkg/commands/chgm/chgm.go:51]: I am not sure if we need this additional struct since ChgmProvider already implements the interface. Lets have a video review/coding sessions tomorrow to do some testing on this.

**Comment by rafael-azevedo** [comment]
@georgettica Just to follow up on the conversation we had earlier we will be able to check the userIdentity element from AWS . On there there is a type that tells us if its an IAMUser or AssumeRole. That will let us know which arn or userName to check when verifying. https://docs.aws.amazon.com/awscloudtrail/latest/userguide/cloudtrail-event-reference-user-identity.html

As far as which users/roles we that are redhat related. I am working on that list right now and will have to sync with a couple of people on it tomorrow as I was not able to today.

File: pkg/commands/chgm/chgm.go (hunk 11 above)
Comment (shibumi) [reply pkg/commands/chgm/chgm.go:129]: You might want to modify the function doc strings. As far as I understand your function it does more than just checking if the instances are running, correct?  You also check in the end if the user is allowed to stop them :thinking:

File: cadctl/cmd/cluster-missing/cluster-missing.go [hunk 18]
```diff
@@ -29,45 +30,68 @@ import (
...
 }
 
 var (
-	incidentID string
+	// externalClusterID has a value although it's the default to denote that this is the default value, if changed it will be the default value for externalClusterID_flagName
+	externalClusterID string = ""
 )
 
 func init() {
-	ClusterMissingCmd.Flags().StringVarP(&incidentID, "payload", "p", "", "The incident payload as received from the PagerDuty WebHook")
```
Comment (shibumi) [inline cadctl/cmd/cluster-missing/cluster-missing.go:72]: @georgettica  why did you remove the incident payload?

Comment (georgettica) [reply cadctl/cmd/cluster-missing/cluster-missing.go:72]: just because I didn't use it now.
currently the entrypoint is externalid, once we write code that ties from payload to externalid I will add it here.

also, it's easier to test when you provide the externalid

Comment (shibumi) [reply cadctl/cmd/cluster-missing/cluster-missing.go:72]: makes sense, thanks for the clarification.

File: pkg/commands/chgm/chgm.go (hunk 11 above)
Comment (georgettica) [reply pkg/commands/chgm/chgm.go:111]: is that good enough?

**Comment by georgettica** [comment]
@rafael-azevedo I added todos, please comment on each of these so we can close them (or even create cards to solve them)

**Comment by georgettica** [comment]
/unhold
this is ready for review @rafael-azevedo

**Comment by georgettica** [comment]
the entrypoint command is not `AreAllInstancesRunningOrCorrectlyStopped`

File: pkg/commands/chgm/chgm.go (hunk 17 above)
Comment (georgettica) [reply pkg/commands/chgm/chgm.go:55]: we missed the 1:1 / coding session on this
hopefully this is the last comment before merge 🤞

**Review by georgettica — COMMENTED** [review]
please @rafael-azevedo when you review, check the TODO's I added.
as this is a large PR and basic logic I think is taken care of, we can address each of the TODOS's and split the work on them (this way all are using the same codebase)

PS, this is the place to decide where to put things before they solidify a bit

File: pkg/commands/isrunning/isrunning.go (hunk 4 above)
Comment (georgettica) [reply pkg/commands/isrunning/isrunning.go]: solved point 3, but the others are not addressed

File: pkg/commands/isrunning/isrunning_test.go (hunk 5 above)
Comment (georgettica) [reply pkg/commands/isrunning/isrunning_test.go]: basically a self review note, in case someone wants to comment against it

File: pkg/commands/chgm/chgm.go (hunk 11 above)
Comment (georgettica) [reply pkg/commands/chgm/chgm.go:110]: @ramonbutter @shibumi feel free to resolve the review thread if you feel your comments have been addressed / codified.
if I didn't. please provide concrete things you would like for me to change / card out to another time

File: cadctl/cmd/cluster-missing/cluster-missing.go (hunk 18 above)
Comment (georgettica) [reply cadctl/cmd/cluster-missing/cluster-missing.go:88]: keeping this open as this might be a question others reviewers might ask

File: pkg/commands/chgm/chgm.go [hunk 19]
```diff
@@ -0,0 +1,224 @@
...
+// having a function change break my code.
+// TODO: decide if the Client should be the ChgmProvider
+type Client struct {
+	ChgmService
+}
+
+// isUserAllowedToStop will verify if a user is allowed to stop instances.
+// as this is the thing that might change the most it is at the top.
+// additionallyit is private as I don't see anyone using this outside of AreAllInstancesRunning
+func isUserAllowedToStop(username string, userDetails CloudTrailEventRaw, infraID string) bool {
```
Comment (rafael-azevedo) [inline pkg/commands/chgm/chgm.go:62]: we can remove the is

File: pkg/commands/chgm/chgm.go [hunk 20]
```diff
@@ -0,0 +1,224 @@
...
+	if strings.HasPrefix(username, "osdManagedAdmin-") {
+		return true
+	}
+
+	return isAssumedRoleOfName("OrganizationAccountAccessRole", userDetails)
+}
+
+// AreAllInstancesRunningOrCorrectlyStopped will check all of the instances under the externalID are running.
+// in case they are not it will make sure the stopped instances are correctly at this state.
+func (i Client) AreAllInstancesRunningOrCorrectlyStopped(externalID string) (bool, error) {
```
Comment (georgettica) [inline pkg/commands/chgm/chgm.go:93]: InvestigateInstances

NonRunningInstances struct{
instances
user
stopTime
userAuthorized
}

File: pkg/commands/chgm/chgm.go [hunk 21]
```diff
@@ -0,0 +1,224 @@
...
+		return CloudTrailEventRaw{}, fmt.Errorf("could not marshal event.CloudTrailEvent: %w", err)
+	}
+	const supportedEventVersion = "1.08"
+	if res.EventVersion != supportedEventVersion {
+		return CloudTrailEventRaw{}, fmt.Errorf("event version differs from saved one (got %s, want %s) , not sure it's the same schema", res.EventVersion, supportedEventVersion)
+	}
+	return res, nil
+}
+
+// isAssumedRoleOfName will check the CloudTrailEventRaw uf the issues is of that role name
```
Comment (georgettica) [inline pkg/commands/chgm/chgm.go:202]:
```suggestion
// isAssumedRoleOfName will verify the SessionIssuer RoleName is the same as the provided
```

File: pkg/commands/chgm/chgm.go [hunk 22]
```diff
@@ -0,0 +1,224 @@
...
+	// if the type is not role, it's not supported for now
+	if userDetails.UserIdentity.SessionContext.SessionIssuer.Type != "Role" {
+		return false
+	}
+	return userDetails.UserIdentity.SessionContext.SessionIssuer.UserName == role
+}
+
+// findMatchingStoppedInstance should extract the correct instance which is related to the provided event
+// TODO: fill this with correct code
+func findMatchingStoppedInstance(event cloudtrail.Event, instances []*ec2.Instance) *ec2.Instance {
```
Comment (georgettica) [inline pkg/commands/chgm/chgm.go:222]: remove

**Comment by georgettica** [comment]
/hold
While writing the code and fixing

File: pkg/commands/chgm/chgm.go (hunk 11 above)
Comment (rafael-azevedo) [reply pkg/commands/chgm/chgm.go:109]: @georgettica I think we talked about this but I forgot to note it down. Since we now have the retry in getting the cloudtrail and we know if instances are stopped we dont need to look at the node count total. At least we can move this into a seperate PR discussion later.

File: pkg/commands/chgm/chgm.go (hunk 22 above)
Comment (rafael-azevedo) [reply pkg/commands/chgm/chgm.go:221]: this still needs to be removed

File: pkg/aws/aws.go [hunk 23]
```diff
@@ -177,8 +183,103 @@ func (a *AwsClient) listInstancesWithFilter(filters []*ec2.Filter) ([]*ec2.Insta
 	return instances, nil
 }
 
+func (a Client) ListStopEventsForSpecificInstances(instances []*ec2.Instance) ([]*cloudtrail.Event, error) {
```
[the remark this answers is not in the document]
Comment (georgettica) [reply pkg/aws/aws.go:186]: Will fix :)

File: pkg/aws/aws.go [hunk 24]
```diff
@@ -177,8 +183,103 @@ func (a *AwsClient) listInstancesWithFilter(filters []*ec2.Filter) ([]*ec2.Insta
...
+		return true, nil
+	})
+
+	if err != nil {
+		return nil, err
+	}
+
+	return events, nil
+}
+func populateIdToStopTimeMap(instances []*ec2.Instance) (map[string]time.Time, error) {
```
[the remark this answers is not in the document]
Comment (georgettica) [reply pkg/aws/aws.go:248]: Ok:)

File: pkg/commands/chgm/chgm.go [hunk 25]
```diff
@@ -0,0 +1,223 @@
...
+	if strings.HasPrefix(username, "osdManagedAdmin-") {
+		return true
+	}
+
+	return assumedRoleOfName("OrganizationAccountAccessRole", userDetails)
+}
+
+// InvestigateInstances will check all of the instances under the externalID are running.
+// in case they are not it will make sure the stopped instances are correctly at this state.
+func (i Client) InvestigateInstances(externalID string) (bool, error) {
```
[the remark this answers is not in the document]
Comment (georgettica) [reply pkg/commands/chgm/chgm.go:92]: No real reason, will fix:)

File: pkg/aws/aws.go [hunk 26]
```diff
@@ -177,8 +183,103 @@ func (a *AwsClient) listInstancesWithFilter(filters []*ec2.Filter) ([]*ec2.Insta
...
+
+	events := []*cloudtrail.Event{}
+
+	idToStopTime, err := populateIdToStopTimeMap(instances)
+	if err != nil {
+		return nil, fmt.Errorf("could not populate the idToCloudtrailEvent map: %w", err)
+	}
+
+	err = wait.ExponentialBackoff(backoffoptions, func() (bool, error) {
+		idToCloudtrailEvent := make(map[string]*cloudtrail.Event)
```
Comment (ramonbutter) [inline pkg/aws/aws.go:202]: I would put this outside of the backoff retry, as we are by allocating new memory on every retry. And not sure, how fast garbage collection will be.

File: pkg/aws/aws.go (hunk 23 above)
[the remark this answers is not in the document]
Comment (georgettica) [reply pkg/aws/aws.go]: chose a different name, but should be ok (🤞 )

File: pkg/commands/chgm/chgm.go [hunk 27]
```diff
@@ -0,0 +1,224 @@
...
+		return true, nil
+	}
+
+	stoppedInstancesEvents, err := i.ListStopInstancesEvents()
+	if err != nil {
+		return false, fmt.Errorf("could not retrieve stopped instance event: %w", err)
+	}
+	//fmt.Printf("stoppedInstancesEvents ::: %#v\n", stoppedInstancesEvents)
+
+	if len(stoppedInstancesEvents) == 0 {
```
[the remark this answers is not in the document]
Comment (georgettica) [reply pkg/commands/chgm/chgm.go]: yes

File: cadctl/cmd/cluster-missing/cluster-missing.go (hunk 18 above)
Comment (georgettica) [reply cadctl/cmd/cluster-missing/cluster-missing.go:98]: wasn't asked :/
not sure if I want to close it though

File: pkg/commands/isrunning/isrunning.go (hunk 4 above)
Comment (georgettica) [reply pkg/commands/isrunning/isrunning.go]: resolved these in the aws package

**Comment by georgettica** [comment]
/unhold
@rafael-azevedo please lgtm
